File: src/main/java/com/itmo/microservices/tasks/service/TaskService.java
package com.itmo.microservices.tasks.service;

import com.itmo.microservices.common.exception.NotFoundException;
import com.itmo.microservices.tasks.entity.Task;
import com.itmo.microservices.tasks.model.TaskDto;
import com.itmo.microservices.tasks.repository.TaskRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@AllArgsConstructor
public class TaskService {

    private final TaskRepository taskRepository;

    public List<Task> getAllTasks() {
        return taskRepository.findAll();
    }

    public Task getTaskById(UUID taskId) {
        return taskRepository.findById(taskId)
                .orElseThrow(() -> new NotFoundException("Task " + taskId + " not found"));
    }

    public void addTask(TaskDto task) {
        taskRepository.save(
                Task.builder()
                        .author(task.getAuthor())
                        .assignee(task.getAssignee())
                        .description(task.getDescription())
                        .status(task.getStatus())
                        .build()
        );
    }

    public void assignTask(UUID taskId, String username) {
        Task task = taskRepository
                .findById(taskId)
                .orElseThrow(() -> new NotFoundException("Task " + taskId + " not found"));
        task.setAssignee(username);
        taskRepository.save(task);
    }

    public void deleteTaskById(UUID taskId) {
        taskRepository.deleteById(taskId);
    }
}

File: src/main/kotlin/com/itmo/microservices/demo/tasks/service/TaskService.kt
package com.itmo.microservices.demo.tasks.service

import com.itmo.microservices.demo.exception.NotFoundException
import com.itmo.microservices.demo.tasks.entity.Task
import lombok.AllArgsConstructor
import com.itmo.microservices.demo.tasks.repository.TaskRepository
import java.util.UUID
import com.itmo.microservices.demo.tasks.model.TaskDto
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
@AllArgsConstructor
class TaskService(private val taskRepository: TaskRepository) {

    fun allTasks(): List<Task> = taskRepository.findAll().filterNotNull()

    fun getTaskById(taskId: UUID): Task {
        return taskRepository.findByIdOrNull(taskId) ?: throw NotFoundException("Task $taskId not found")
    }

    fun addTask(task: TaskDto) {
        //TODO 04.08.2021 Ilya_Kuznetsov: pass author as AuthenticationPrincipal from controller
        val entity = Task(author=task.author, assignee=task.assignee, title = task.title, description = task.description, status = task.status)
        taskRepository.save(entity)
    }

    fun assignTask(taskId: UUID, username: String) {
        val task = taskRepository.findByIdOrNull(taskId) ?: throw NotFoundException("Task $taskId not found")
        task.assignee = username
        taskRepository.save(task)
    }

    fun deleteTaskById(taskId: UUID) = taskRepository.deleteById(taskId)
}